Integrate requestIdleCallback with HTML5 event loop and OpportunisticTaskScheduler

rniwa opened the pull request:
#### 03c76b1a950a2302989856fd66a4baa6fa6914cd
<pre>
Integrate requestIdleCallback with HTML5 event loop and OpportunisticTaskScheduler
<a href="https://bugs.webkit.org/show_bug.cgi?id=259850">https://bugs.webkit.org/show_bug.cgi?id=259850</a>

Reviewed by Yusuke Suzuki.

This PR integrates WebKit&apos;s implementation of requestIdleCallback with HTML5 event loop and Page&apos;s
OpportunisticTaskScheduler. It also skips the tests on non-CF (non-Apple) ports as RunLoopObserver,
which is used by OpportunisticTaskScheduler isn&apos;t implemented on non-CF ports.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-idle-periods-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-multiple-calls-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-xhr-sync-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache-expected.txt: Updated the test to not
test the order of execution of tasks scheduled across distinct frames as we no longer have any guarantee
across them (as spec&apos;ed).
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache.html: Ditto.
* LayoutTests/requestidlecallback/requestidlecallback-is-not-called-when-canceled-expected.txt: Ditto.
* LayoutTests/requestidlecallback/requestidlecallback-is-not-called-when-canceled.html: Ditto.

* Source/WebCore/dom/EventLoop.h:
(WebCore::EventLoop::hasTasksForFullyActiveDocument const): Added.

* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback):
(WebCore::IdleCallbackController::startIdlePeriod):

* Source/WebCore/dom/IdleCallbackController.h:

* Source/WebCore/dom/IdleDeadline.cpp:
(WebCore::IdleDeadline::didTimeout const): Deleted.

* Source/WebCore/dom/IdleDeadline.h:
(WebCore::IdleDeadline::create): Now takes DidTimeout enum as an argument.
(WebCore::IdleDeadline::didTimeout const): Moved from cpp file. Now simply checks DidTimeout given to
the constructor.

* Source/WebCore/dom/Microtasks.h:
(WebCore::MicrotaskQueue::isEmpty): Added.

* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks): Added.
(WebCore::WindowEventLoop::shouldEndIdlePeriod): Added.

* Source/WebCore/dom/WindowEventLoop.h:
(WebCore::WindowEventLoop::didScheduleRenderingUpdate): Added.
(WebCore::WindowEventLoop::didFinishRenderingUpdate): Added.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal): Added hooks for WindowEventLoop. This function is
used instead of didScheduleRenderingUpdate since the latter isn&apos;t called when the chrome client
scheduled the rendering update.
(WebCore::Page::doAfterUpdateRendering): Ditto.
(WebCore::Page::renderingUpdateCompleted):
(WebCore::Page::forEachWindowEventLoop): Added.
(WebCore::Page::performOpportunisticallyScheduledTasks): Call WindowEventLoop&apos;s
opportunisticallyRunIdleCallbacks to invoke idle callbacks.

* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/266622@main">https://commits.webkit.org/266622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd706c8860591a10efa36380ef8aa7635b98f597

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16048 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14433 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14698 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16214 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15036 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16766 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12321 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12898 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19910 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13398 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13062 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16272 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11462 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12901 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12758 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17237 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1706 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13459 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3841,6 +3851,27 @@ void Page::forEachFrame(const Function<void(LocalFrame&)>& functor)
+        lastEventLoop = currentEventLoop.ptr();
+        windowEventLoops.add(WTFMove(currentEventLoop));
+    }
+    for (auto& eventLoop : windowEventLoops)
+        functor(eventLoop);
```
OK, this means we do not have strict ordering of these even loops (since HashSet), is it correct?

rniwa replied:
Yes. As far as I've read, this is per spec.

Constellation commented on Source/WebCore/dom/WindowEventLoop.h:
```diff
@@ -53,6 +53,12 @@ class WindowEventLoop final : public EventLoop {
 
     CustomElementQueue& backupElementQueue();
 
+    void didScheduleRenderingUpdate() { m_hasARenderingOpportunity = true; }
+    void didFinishRenderingUpdate() { m_hasARenderingOpportunity = false; }
```
So we do not run idle callback when next rendering is already scheduled (and waiting), makes sense.
Is it better to ask to RenderingUpdateScheduler directly? (For example, taking `Page&` in `opportunisticallyRunIdleCallbacks`, threading it to `shouldEndIdlePeriod` and ask to `RenderingUpdateScheduler` instead of configuring this bool in each WindowEventLoop.

rniwa replied:
So one subtle thing here is that we don't use Page's `RenderingUpdateScheduler` to schedule rendering update in WebKit2. Instead, ChromeClient in WebKit2 schedules rendering update and we early exit in `scheduleRenderingUpdateInternal`.

Constellation commented on Source/WebCore/dom/IdleCallbackController.cpp:
```diff
@@ -93,7 +87,9 @@ void IdleCallbackController::startIdlePeriod()
         m_runnableIdleCallbacks.append({ request.identifier, WTFMove(request.callback) });
```
For deadline, right now, we are computing it per `IdleCallbackController`, but this means that if we have multiple documents, each document can run 50ms of tasks. Should we have a global deadline instead? (or per window?, I think the current spec text is per-window (though getDeadline implementation allows it to be more global one)).

rniwa replied:
Right, I have a WIP PR to do that as a follow up.

rniwa replied:
New PR up at https://github.com/WebKit/WebKit/pull/16426

Constellation commented on Source/WebCore/dom/WindowEventLoop.h:
```diff
@@ -82,6 +88,10 @@ class WindowEventLoop final : public EventLoop {
+
+    bool m_hasARenderingOpportunity { false };
+    MonotonicTime m_lastIdlePeriodStartTime;
+    MonotonicTime m_nextRenderingUpdateTime;
```
What is the purpose of these two times?

rniwa replied:
Oh oops, these are useless in this version. Will remove.

Constellation reviewed (approved):
r=me with some comments.

rniwa committed:
Integrate requestIdleCallback with HTML5 event loop and OpportunisticTaskScheduler
https://bugs.webkit.org/show_bug.cgi?id=259850

Reviewed by Yusuke Suzuki.

This PR integrates WebKit's implementation of requestIdleCallback with HTML5 event loop and Page's
OpportunisticTaskScheduler. It also skips the tests on non-CF (non-Apple) ports as RunLoopObserver,
which is used by OpportunisticTaskScheduler isn't implemented on non-CF ports.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-idle-periods-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-multiple-calls-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/callback-xhr-sync-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache-expected.txt: Updated the test to not
test the order of execution of tasks scheduled across distinct frames as we no longer have any guarantee
across them (as spec'ed).
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache.html: Ditto.
* LayoutTests/requestidlecallback/requestidlecallback-is-not-called-when-canceled-expected.txt: Ditto.
* LayoutTests/requestidlecallback/requestidlecallback-is-not-called-when-canceled.html: Ditto.

* Source/WebCore/dom/EventLoop.h:
(WebCore::EventLoop::hasTasksForFullyActiveDocument const): Added.

* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback):
(WebCore::IdleCallbackController::startIdlePeriod):

* Source/WebCore/dom/IdleCallbackController.h:

* Source/WebCore/dom/IdleDeadline.cpp:
(WebCore::IdleDeadline::didTimeout const): Deleted.

* Source/WebCore/dom/IdleDeadline.h:
(WebCore::IdleDeadline::create): Now takes DidTimeout enum as an argument.
(WebCore::IdleDeadline::didTimeout const): Moved from cpp file. Now simply checks DidTimeout given to
the constructor.

* Source/WebCore/dom/Microtasks.h:
(WebCore::MicrotaskQueue::isEmpty): Added.

* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks): Added.
(WebCore::WindowEventLoop::shouldEndIdlePeriod): Added.

* Source/WebCore/dom/WindowEventLoop.h:
(WebCore::WindowEventLoop::didScheduleRenderingUpdate): Added.
(WebCore::WindowEventLoop::didFinishRenderingUpdate): Added.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal): Added hooks for WindowEventLoop. This function is
used instead of didScheduleRenderingUpdate since the latter isn't called when the chrome client
scheduled the rendering update.
(WebCore::Page::doAfterUpdateRendering): Ditto.
(WebCore::Page::renderingUpdateCompleted):
(WebCore::Page::forEachWindowEventLoop): Added.
(WebCore::Page::performOpportunisticallyScheduledTasks): Call WindowEventLoop's
opportunisticallyRunIdleCallbacks to invoke idle callbacks.

* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/266622@main
